File: server/python/src/infrastructure/persistence/sqlite_item_repository.py
"""SQLite implementation of Item repository."""

import sqlite3
from datetime import datetime, timezone
from pathlib import Path

from ...domain.entities.item import (
    Item,
    CreateItemData,
    UpdateItemData,
    PatchItemData,
    item_from_row,
    serialize_images_column,
)
from ...domain.repositories.item_repository import (
    ItemRepository,
    FilterOptions,
    SortOptions,
    SortField,
    SortDirection,
    PaginationOptions,
    SearchOptions,
    PaginatedResult,
)
from ...shared.config import config
from ...shared.cursor import encode_cursor, decode_cursor, CursorData


class SQLiteItemRepository(ItemRepository):
    """SQLite repository adapter for items."""

    def __init__(self, db_path: Path | None = None) -> None:
        """
        Initialize the repository.

        Args:
            db_path: Path to SQLite database. If None, uses config.database_path
        """
        self._db_path = db_path or config.database_path

    def _get_connection(self) -> sqlite3.Connection:
        """
        Create a new database connection.

        Returns:
            SQLite connection with row factory configured
        """
        conn = sqlite3.connect(str(self._db_path))
        conn.row_factory = sqlite3.Row
        return conn

    async def create(self, data: CreateItemData) -> Item:
        """Create a new item."""
        now = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")

        images_json = serialize_images_column(data.images or [])

        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                INSERT INTO items (
                    title, description, price_cents, category, condition, status,
                    is_featured, city, postal_code, country, delivery_available,
                    images, created_at, updated_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    data.title,
                    data.description,
                    data.price_cents,
                    data.category,
                    data.condition.value,
                    data.status.value,
                    1 if data.is_featured else 0,
                    data.city,
                    data.postal_code,
                    data.country,
                    1 if data.delivery_available else 0,
                    images_json,
                    now,
                    now,
                ),
            )
            conn.commit()
            item_id = cursor.lastrowid

        created = await self.find_by_id(item_id)
        if created is None:
            raise RuntimeError("Failed to create item")
        return created

    async def find_by_id(self, id: int) -> Item | None:
        """Find an item by ID."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute("SELECT * FROM items WHERE id = ?", (id,))
            row = cursor.fetchone()

            if row is None:
                return None

            return item_from_row(dict(row))

    async def find_all(
        self,
        filters: FilterOptions | None = None,
        sort: SortOptions | None = None,
        pagination: PaginationOptions | None = None,
    ) -> PaginatedResult:
        """List all items with filtering, sorting, and pagination."""
        where_conditions: list[str] = []
        params: list = []

        if filters:
            if filters.status:
                where_conditions.append("status = ?")
                params.append(filters.status)
            if filters.category:
                where_conditions.append("category = ?")
                params.append(filters.category)
            if filters.city:
                where_conditions.append("city = ?")
                params.append(filters.city)
            if filters.postal_code:
                where_conditions.append("postal_code = ?")
                params.append(filters.postal_code)
            if filters.is_featured is not None:
                where_conditions.append("is_featured = ?")
                params.append(1 if filters.is_featured else 0)
            if filters.delivery_available is not None:
                where_conditions.append("delivery_available = ?")
                params.append(1 if filters.delivery_available else 0)

        # Handle cursor pagination
        if pagination and pagination.cursor:
            cursor_data = decode_cursor(pagination.cursor)
            if cursor_data:
                sort_field = sort.field if sort else SortField.CREATED_AT
                sort_dir = sort.direction if sort else SortDirection.DESC

                if sort_dir == SortDirection.DESC:
                    where_conditions.append(f"({sort_field.value}, id) < (?, ?)")
                else:
                    where_conditions.append(f"({sort_field.value}, id) > (?, ?)")
                params.extend([cursor_data.created_at, cursor_data.id])

        where_clause = f"WHERE {' AND '.join(where_conditions)}" if where_conditions else ""

        # Build ORDER BY clause
        sort_field = sort.field if sort else SortField.CREATED_AT
        sort_dir = sort.direction if sort else SortDirection.DESC
        order_clause = f"ORDER BY {sort_field.value} {sort_dir.value.upper()}, id {sort_dir.value.upper()}"

        # Build LIMIT clause
        limit = pagination.limit if pagination else 20
        limit_clause = f"LIMIT {limit + 1}"

        # Execute query
        query = f"SELECT * FROM items {where_clause} {order_clause} {limit_clause}"
        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(query, params)
            rows = cursor.fetchall()

        # Check for next page
        has_next_page = len(rows) > limit
        items = [item_from_row(dict(row)) for row in rows[:limit]]

        # Build next cursor
        next_cursor: str | None = None
        if has_next_page and items:
            last_item = items[-1]
            cursor_data = CursorData(id=last_item.id, created_at=last_item.created_at)
            next_cursor = encode_cursor(cursor_data)

        return PaginatedResult(items=items, next_cursor=next_cursor)

    async def search(self, options: SearchOptions) -> PaginatedResult:
        """Full-text search using FTS5."""
        where_conditions: list[str] = []
        params: list = []

        if options.filters:
            if options.filters.status:
                where_conditions.append("items.status = ?")
                params.append(options.filters.status)
            if options.filters.category:
                where_conditions.append("items.category = ?")
                params.append(options.filters.category)
            if options.filters.city:
                where_conditions.append("items.city = ?")
                params.append(options.filters.city)
            if options.filters.postal_code:
                where_conditions.append("items.postal_code = ?")
                params.append(options.filters.postal_code)
            if options.filters.is_featured is not None:
                where_conditions.append("items.is_featured = ?")
                params.append(1 if options.filters.is_featured else 0)
            if options.filters.delivery_available is not None:
                where_conditions.append("items.delivery_available = ?")
                params.append(1 if options.filters.delivery_available else 0)

        # Handle cursor pagination
        if options.pagination and options.pagination.cursor:
            cursor_data = decode_cursor(options.pagination.cursor)
            if cursor_data:
                where_conditions.append("(items.created_at, items.id) < (?, ?)")
                params.extend([cursor_data.created_at, cursor_data.id])

        filters_clause = f"AND {' AND '.join(where_conditions)}" if where_conditions else ""

        # Build LIMIT clause
        limit = options.pagination.limit if options.pagination else 20
        limit_clause = f"LIMIT {limit + 1}"

        # Escape FTS search query
        fts_query = options.query.replace('"', '""')

        # Execute FTS search with BM25 ranking
        search_query = f"""
            SELECT items.* FROM items
            INNER JOIN items_fts ON items.id = items_fts.rowid
            WHERE items_fts MATCH ?
            {filters_clause}
            ORDER BY bm25(items_fts), items.created_at DESC, items.id DESC
            {limit_clause}
        """

        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(search_query, [fts_query] + params)
            rows = cursor.fetchall()

        # Check for next page
        has_next_page = len(rows) > limit
        items = [item_from_row(dict(row)) for row in rows[:limit]]

        # Build next cursor
        next_cursor: str | None = None
        if has_next_page and items:
            last_item = items[-1]
            cursor_data = CursorData(id=last_item.id, created_at=last_item.created_at)
            next_cursor = encode_cursor(cursor_data)

        return PaginatedResult(items=items, next_cursor=next_cursor)

    async def update(self, id: int, data: UpdateItemData) -> Item | None:
        """Update (replace) an item."""
        now = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")

        images_json = serialize_images_column(data.images or [])

        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(
                """
                UPDATE items SET
                    title = ?, description = ?, price_cents = ?, category = ?,
                    condition = ?, status = ?, is_featured = ?, city = ?,
                    postal_code = ?, country = ?, delivery_available = ?,
                    images = ?, updated_at = ?
                WHERE id = ?
                """,
                (
                    data.title,
                    data.description,
                    data.price_cents,
                    data.category,
                    data.condition.value,
                    data.status.value,
                    1 if data.is_featured else 0,
                    data.city,
                    data.postal_code,
                    data.country,
                    1 if data.delivery_available else 0,
                    images_json,
                    now,
                    id,
                ),
            )
            conn.commit()

            if cursor.rowcount == 0:
                return None

        return await self.find_by_id(id)

    async def patch(self, id: int, data: PatchItemData) -> Item | None:
        """Partially update an item."""
        now = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")

        updates: list[str] = []
        params: list = []

        if data.title is not None:
            updates.append("title = ?")
            params.append(data.title)
        if data.description is not None:
            updates.append("description = ?")
            params.append(data.description)
        if data.price_cents is not None:
            updates.append("price_cents = ?")
            params.append(data.price_cents)
        if data.category is not None:
            updates.append("category = ?")
            params.append(data.category)
        if data.condition is not None:
            updates.append("condition = ?")
            params.append(data.condition.value)
        if data.status is not None:
            updates.append("status = ?")
            params.append(data.status.value)
        if data.is_featured is not None:
            updates.append("is_featured = ?")
            params.append(1 if data.is_featured else 0)
        if data.city is not None:
            updates.append("city = ?")
            params.append(data.city)
        if data.postal_code is not None:
            updates.append("postal_code = ?")
            params.append(data.postal_code)
        if data.country is not None:
            updates.append("country = ?")
            params.append(data.country)
        if data.delivery_available is not None:
            updates.append("delivery_available = ?")
            params.append(1 if data.delivery_available else 0)
        if data.images is not None:
            updates.append("images = ?")
            params.append(serialize_images_column(data.images))

        updates.append("updated_at = ?")
        params.append(now)
        params.append(id)

        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute(f"UPDATE items SET {', '.join(updates)} WHERE id = ?", params)
            conn.commit()

            if cursor.rowcount == 0:
                return None

        return await self.find_by_id(id)

    async def delete(self, id: int) -> bool:
        """Delete an item."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute("DELETE FROM items WHERE id = ?", (id,))
            conn.commit()
            return cursor.rowcount > 0

    async def exists(self, id: int) -> bool:
        """Check if an item exists."""
        with self._get_connection() as conn:
            cursor = conn.cursor()
            cursor.execute("SELECT 1 FROM items WHERE id = ? LIMIT 1", (id,))
            return cursor.fetchone() is not None

File: server/kotlin/src/main/kotlin/com/lebonpoint/infrastructure/persistence/SQLiteItemRepository.kt
package com.lebonpoint.infrastructure.persistence

import com.lebonpoint.domain.entities.CreateItemData
import com.lebonpoint.domain.entities.Item
import com.lebonpoint.domain.entities.PatchItemData
import com.lebonpoint.domain.entities.ReplaceItemData
import com.lebonpoint.domain.entities.applyPatch
import com.lebonpoint.domain.repositories.ItemFilters
import com.lebonpoint.domain.repositories.ItemPage
import com.lebonpoint.domain.repositories.ItemRepository
import com.lebonpoint.domain.repositories.SortDirection
import com.lebonpoint.domain.repositories.SortOptions
import com.lebonpoint.domain.valueobjects.ItemCondition
import com.lebonpoint.domain.valueobjects.ItemStatus
import com.lebonpoint.domain.valueobjects.toItemCondition
import com.lebonpoint.domain.valueobjects.toItemStatus
import com.lebonpoint.domain.valueobjects.toDatabaseString
import com.lebonpoint.domain.valueobjects.serializeToJson
import com.lebonpoint.domain.valueobjects.deserializeToImages
import com.lebonpoint.shared.decodeCursor
import com.lebonpoint.shared.encodeCursor
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.sql.Connection
import java.sql.Types
import java.time.Instant
import java.time.format.DateTimeFormatter

/**
 * SQLite implementation of ItemRepository
 * Uses JDBC with connection pooling via HikariCP
 */
class SQLiteItemRepository(
    private val dbConfig: DatabaseConfig = DatabaseConfig
) : ItemRepository {

    companion object {
        private val ISO_FORMATTER = DateTimeFormatter.ISO_OFFSET_DATE_TIME

        /**
         * Parse ISO 8601 timestamp string to Instant
         */
        fun parseInstant(value: String?): Instant? {
            return value?.let {
                try {
                    Instant.parse(it)
                } catch (e: Exception) {
                    null
                }
            }
        }

        /**
         * Format Instant to ISO 8601 string
         */
        fun formatInstant(instant: Instant?): String? {
            return instant?.toString()
        }
    }

    override suspend fun create(data: CreateItemData): Item = withContext(Dispatchers.IO) {
        dbConfig.withConnection { conn ->
            val now = Instant.now()

            conn.prepareStatement("""
                INSERT INTO items (
                    title, description, price_cents, category, condition, status,
                    is_featured, city, postal_code, country, delivery_available,
                    created_at, updated_at, published_at, images
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """.trimIndent()).use { stmt ->
                stmt.setString(1, data.title)
                stmt.setString(2, data.description)
                stmt.setInt(3, data.priceCents)
                stmt.setString(4, data.category)
                stmt.setString(5, data.condition.toDatabaseString())
                stmt.setString(6, data.status.toDatabaseString())
                stmt.setInt(7, if (data.isFeatured) 1 else 0)
                stmt.setString(8, data.city)
                stmt.setString(9, data.postalCode)
                stmt.setString(10, data.country)
                stmt.setInt(11, if (data.deliveryAvailable) 1 else 0)
                stmt.setString(12, formatInstant(now))
                stmt.setString(13, formatInstant(now))
                stmt.setString(14, null) // published_at
                stmt.setString(15, data.images.serializeToJson())

                stmt.executeUpdate()

                val generatedId = conn.createStatement().use { keyStmt ->
                    keyStmt.executeQuery("SELECT last_insert_rowid()").use { rs ->
                        if (rs.next()) {
                            rs.getInt(1)
                        } else {
                            throw IllegalStateException("Failed to get generated ID")
                        }
                    }
                }

                upsertFullTextIndex(conn = conn, id = generatedId, title = data.title, description = data.description)

                Item(
                    id = generatedId,
                    title = data.title,
                    description = data.description,
                    priceCents = data.priceCents,
                    category = data.category,
                    condition = data.condition,
                    status = data.status,
                    isFeatured = data.isFeatured,
                    city = data.city,
                    postalCode = data.postalCode,
                    country = data.country,
                    deliveryAvailable = data.deliveryAvailable,
                    createdAt = now,
                    updatedAt = now,
                    publishedAt = null,
                    images = data.images
                )
            }
        }
    }

    override suspend fun findById(id: Int): Item? = withContext(Dispatchers.IO) {
        dbConfig.withConnection { conn ->
            conn.prepareStatement("SELECT * FROM items WHERE id = ?").use { stmt ->
                stmt.setInt(1, id)
                stmt.executeQuery().use { rs ->
                    if (rs.next()) {
                        mapRowToItem(rs)
                    } else {
                        null
                    }
                }
            }
        }
    }

    override suspend fun findAll(
        filters: ItemFilters,
        sort: SortOptions,
        limit: Int,
        cursor: String?
    ): ItemPage = withContext(Dispatchers.IO) {
        dbConfig.withConnection { conn ->
            val whereClause = buildWhereClause(filters)
            val orderByClause = buildOrderByClause(sort)
            val cursorFilter = buildCursorFilter(cursor)

            val combinedWhere = listOfNotNull(
                whereClause,
                cursorFilter
            ).let { clauses ->
                if (clauses.isNotEmpty()) "WHERE ${clauses.joinToString(" AND ")}" else ""
            }

            val sql = """
                SELECT * FROM items
                $combinedWhere
                $orderByClause
                LIMIT ?
            """.trimIndent()

            conn.prepareStatement(sql).use { stmt ->
                var paramIndex = 1
                paramIndex = setFilterParameters(stmt, filters, paramIndex)
                stmt.setInt(paramIndex, limit.coerceIn(1, 100))

                stmt.executeQuery().use { rs ->
                    val items = mutableListOf<Item>()
                    while (rs.next()) {
                        items.add(mapRowToItem(rs))
                    }

                    val nextCursor = items.lastOrNull()?.let {
                        if (items.size == limit) encodeCursor(it.id) else null
                    }

                    ItemPage(items, nextCursor)
                }
            }
        }
    }

    override suspend fun search(
        query: String,
        filters: ItemFilters,
        sort: SortOptions,
        limit: Int,
        cursor: String?
    ): ItemPage = withContext(Dispatchers.IO) {
        dbConfig.withConnection { conn ->
            val whereClause = buildWhereClause(filters)
            val cursorFilter = buildCursorFilter(cursor)

            val combinedWhere = listOfNotNull(
                "items_fts MATCH ?",
                whereClause,
                cursorFilter
            ).let { clauses ->
                "WHERE ${clauses.joinToString(" AND ")}"
            }

            // Use BM25 ranking for relevance
            val orderByClause = if (sort.field == "created_at" && sort.direction == SortDirection.DESC) {
                // Keep default sort by relevance
                "ORDER BY bm25(items_fts) ASC, items.id ASC"
            } else {
                buildOrderByClause(sort)
            }

            val sql = """
                SELECT items.* FROM items
                INNER JOIN items_fts ON items.id = items_fts.rowid
                $combinedWhere
                $orderByClause
                LIMIT ?
            """.trimIndent()

            conn.prepareStatement(sql).use { stmt ->
                var paramIndex = 1
                stmt.setString(paramIndex++, query) // FTS search term
                paramIndex = setFilterParameters(stmt, filters, paramIndex)
                stmt.setInt(paramIndex, limit.coerceIn(1, 100))

                stmt.executeQuery().use { rs ->
                    val items = mutableListOf<Item>()
                    while (rs.next()) {
                        items.add(mapRowToItem(rs))
                    }

                    val nextCursor = items.lastOrNull()?.let {
                        if (items.size == limit) encodeCursor(it.id) else null
                    }

                    ItemPage(items, nextCursor)
                }
            }
        }
    }

    override suspend fun update(id: Int, data: ReplaceItemData): Item? = withContext(Dispatchers.IO) {
        dbConfig.withConnection { conn ->
            val now = Instant.now()

            conn.prepareStatement("""
                UPDATE items SET
                    title = ?, description = ?, price_cents = ?, category = ?, condition = ?,
                    status = ?, is_featured = ?, city = ?, postal_code = ?, country = ?,
                    delivery_available = ?, updated_at = ?, images = ?
                WHERE id = ?
            """.trimIndent()).use { stmt ->
                stmt.setString(1, data.title)
                stmt.setString(2, data.description)
                stmt.setInt(3, data.priceCents)
                stmt.setString(4, data.category)
                stmt.setString(5, data.condition.toDatabaseString())
                stmt.setString(6, data.status.toDatabaseString())
                stmt.setInt(7, if (data.isFeatured) 1 else 0)
                stmt.setString(8, data.city)
                stmt.setString(9, data.postalCode)
                stmt.setString(10, data.country)
                stmt.setInt(11, if (data.deliveryAvailable) 1 else 0)
                stmt.setString(12, formatInstant(now))
                stmt.setString(13, data.images.serializeToJson())
                stmt.setInt(14, id)

                val rows = stmt.executeUpdate()
                if (rows > 0) {
                    upsertFullTextIndex(conn = conn, id = id, title = data.title, description = data.description)
                    conn.prepareStatement("SELECT * FROM items WHERE id = ?").use { selectStmt ->
                        selectStmt.setInt(1, id)
                        selectStmt.executeQuery().use { rs ->
                            if (rs.next()) {
                                mapRowToItem(rs)
                            } else {
                                null
                            }
                        }
                    }
                } else {
                    null
                }
            }
        }
    }

    override suspend fun patch(id: Int, data: PatchItemData): Item? = withContext(Dispatchers.IO) {
        val existing = findById(id) ?: return@withContext null

        val updated = existing.applyPatch(data)

        dbConfig.withConnection { conn ->
            val setClauses = mutableListOf<String>()
            val values = mutableListOf<Any?>()

            data.title?.let { setClauses.add("title = ?"); values.add(it) }
            data.description?.let { setClauses.add("description = ?"); values.add(it) }
            data.priceCents?.let { setClauses.add("price_cents = ?"); values.add(it) }
            data.category?.let { setClauses.add("category = ?"); values.add(it) }
            data.condition?.let { setClauses.add("condition = ?"); values.add(it.toDatabaseString()) }
            data.status?.let { setClauses.add("status = ?"); values.add(it.toDatabaseString()) }
            data.isFeatured?.let { setClauses.add("is_featured = ?"); values.add(if (it) 1 else 0) }
            data.city?.let { setClauses.add("city = ?"); values.add(it) }
            data.postalCode?.let { setClauses.add("postal_code = ?"); values.add(it) }
            data.country?.let { setClauses.add("country = ?"); values.add(it) }
            data.deliveryAvailable?.let { setClauses.add("delivery_available = ?"); values.add(if (it) 1 else 0) }
            data.images?.let { setClauses.add("images = ?"); values.add(it.serializeToJson()) }

            setClauses.add("updated_at = ?")
            values.add(formatInstant(Instant.now()))

            if (setClauses.isEmpty()) {
                return@withConnection existing
            }

            val sql = "UPDATE items SET ${setClauses.joinToString(", ")} WHERE id = ?"
            values.add(id)

            conn.prepareStatement(sql).use { stmt ->
                values.forEachIndexed { index, value ->
                    when (value) {
                        is String -> stmt.setString(index + 1, value)
                        is Int -> stmt.setInt(index + 1, value)
                        null -> stmt.setNull(index + 1, java.sql.Types.VARCHAR)
                    }
                }

                stmt.executeUpdate()
                upsertFullTextIndex(conn = conn, id = id, title = updated.title, description = updated.description)
            }

            updated
        }
    }

    override suspend fun delete(id: Int): Boolean = withContext(Dispatchers.IO) {
        dbConfig.withConnection { conn ->
            conn.prepareStatement("DELETE FROM items WHERE id = ?").use { stmt ->
                stmt.setInt(1, id)
                val deleted = stmt.executeUpdate() > 0
                if (deleted) {
                    deleteFullTextIndex(conn, id)
                }
                deleted
            }
        }
    }

    /**
     * Map database row to Item entity
     */
    private fun mapRowToItem(rs: java.sql.ResultSet): Item {
        return Item(
            id = rs.getInt("id"),
            title = rs.getString("title"),
            description = rs.getString("description"),
            priceCents = rs.getInt("price_cents"),
            category = rs.getString("category"),
            condition = rs.getString("condition").toItemCondition() ?: ItemCondition.UNKNOWN,
            status = rs.getString("status").toItemStatus() ?: ItemStatus.DRAFT,
            isFeatured = rs.getInt("is_featured") == 1,
            city = rs.getString("city"),
            postalCode = rs.getString("postal_code"),
            country = rs.getString("country"),
            deliveryAvailable = rs.getInt("delivery_available") == 1,
            createdAt = parseInstant(rs.getString("created_at")) ?: Instant.EPOCH,
            updatedAt = parseInstant(rs.getString("updated_at")) ?: Instant.EPOCH,
            publishedAt = parseInstant(rs.getString("published_at")),
            images = rs.getString("images").deserializeToImages()
        )
    }

    /**
     * Build WHERE clause from filters
     */
    private fun buildWhereClause(filters: ItemFilters): String? {
        val conditions = mutableListOf<String>()

        filters.status?.let { conditions.add("status = ?") }
        filters.category?.let { conditions.add("category = ?") }
        filters.city?.let { conditions.add("city = ?") }
        filters.postalCode?.let { conditions.add("postal_code = ?") }
        filters.isFeatured?.let { conditions.add("is_featured = ?") }
        filters.deliveryAvailable?.let { conditions.add("delivery_available = ?") }

        return conditions.takeIf { it.isNotEmpty() }?.joinToString(" AND ")
    }

    /**
     * Build ORDER BY clause from sort options
     */
    private fun buildOrderByClause(sort: SortOptions): String {
        val direction = if (sort.direction == SortDirection.ASC) "ASC" else "DESC"
        return "ORDER BY ${sort.field} $direction"
    }

    /**
     * Build cursor filter for pagination
     */
    private fun buildCursorFilter(cursor: String?): String? {
        val cursorData = decodeCursor(cursor) ?: return null
        return "id > ${cursorData.id}"
    }

    /**
     * Set filter parameters on prepared statement
     */
    private fun setFilterParameters(
        stmt: java.sql.PreparedStatement,
        filters: ItemFilters,
        startIndex: Int
    ): Int {
        var index = startIndex

        filters.status?.let { stmt.setString(index++, it) }
        filters.category?.let { stmt.setString(index++, it) }
        filters.city?.let { stmt.setString(index++, it) }
        filters.postalCode?.let { stmt.setString(index++, it) }
        filters.isFeatured?.let { stmt.setInt(index++, if (it) 1 else 0) }
        filters.deliveryAvailable?.let { stmt.setInt(index++, if (it) 1 else 0) }

        return index
    }

    private fun upsertFullTextIndex(
        conn: Connection,
        id: Int,
        title: String?,
        description: String?
    ) {
        conn.prepareStatement("""
            INSERT OR REPLACE INTO items_fts(rowid, title, description)
            VALUES (?, ?, ?)
        """.trimIndent()).use { stmt ->
            stmt.setInt(1, id)
            stmt.setString(2, title)
            if (description != null) {
                stmt.setString(3, description)
            } else {
                stmt.setNull(3, Types.VARCHAR)
            }
            stmt.executeUpdate()
        }
    }

    private fun deleteFullTextIndex(conn: Connection, id: Int) {
        conn.prepareStatement("DELETE FROM items_fts WHERE rowid = ?").use { stmt ->
            stmt.setInt(1, id)
            stmt.executeUpdate()
        }
    }
}

/**
 * Helper function for non-null list of strings
 */
private fun <T> listOfNotNull(vararg elements: T?): List<T> {
    return elements.filterNotNull().toList()
}
